js/background.js

```diff
@@ -483,84 +499,93 @@
+        data.message.flags & textsecure.protobuf.DataMessage.Flags.PROFILE_KEY_UPDATE
+      );
+      if (isProfileUpdate) {
+        return handleProfileUpdate({ data, confirm, messageDescriptor });
+      }
 
-        return isMessageDuplicate(message).then(function(isDuplicate) {
-            if (isDuplicate) {
-                console.log('Received duplicate message', message.idForLogging());
-                ev.confirm();
-                return;
-            }
+      const message = createMessage(data);
+      const isDuplicate = await isMessageDuplicate(message);
+      if (isDuplicate) {
+        console.log('Received duplicate message', message.idForLogging());
+        return event.confirm();
+      }
 
-            return ConversationController.getOrCreateAndWait(id, type).then(function() {
-                return message.handleDataMessage(data.message, ev.confirm, {
-                    initialLoadComplete: initialLoadComplete
-                });
-            });
-        });
-    }
+      const upgradedMessage = await Message.upgradeSchema(data.message);
+      await ConversationController.getOrCreateAndWait(
+        messageDescriptor.id, messageDescriptor.type
+      );
+      return message.handleDataMessage(
+        upgradedMessage,
+        event.confirm,
+        { initialLoadComplete }
+      );
+    };
+  }
+
+  // Received:
+  async function handleMessageReceivedProfileUpdate(
+    { data, confirm, messageDescriptor }
+  ) {
+    const profileKey = data.message.profileKey.toArrayBuffer();
+    const sender = await ConversationController.getOrCreateAndWait(
+      messageDescriptor.source, 'private'
+    );
+    await sender.setProfileKey(profileKey);
+    return confirm();
+  }
+
+  const onMessageReceived = createMessageHandler({
+    handleProfileUpdate: handleMessageReceivedProfileUpdate,
+    getMessageDescriptor: Message.getDescriptorForReceived,
+    createMessage: initIncomingMessage,
+  });
+
+  // Sent:
+  async function handleMessageSentProfileUpdate(
+    { confirm, messageDescriptor }
+  ) {
+    const conversation = await ConversationController.getOrCreateAndWait(
+      messageDescriptor.id, messageDescriptor.type
+    );
+    await conversation.save({ profileSharing: true });
+    return confirm();
+  }
+
+  function createSentMessage(data) {
+    const now = Date.now();
+    return new Whisper.Message({
+      source: textsecure.storage.user.getNumber(),
+      sourceDevice: data.device,
+      sent_at: data.timestamp,
+      received_at: now,
+      conversationId: data.destination,
+      type: 'outgoing',
+      sent: true,
+      expirationStartTimestamp: data.expirationStartTimestamp,
+    });
+  }
+
+  const onSentMessage = createMessageHandler({
+    handleProfileUpdate: handleMessageSentProfileUpdate,
+    getMessageDescriptor: Message.getDescriptorForSent,
+    createMessage: createSentMessage,
+  });
+  /* jshint ignore:end */
+  /* eslint-disable */
```

Comment:
The above are the refactored `MessageReceiver` handlers to allow a single code path that includes message schema upgrades and avoids bloating `handleDataMessage` even more. I’ve tested them manually using both private messages and groups but I’d appreciate a good review of them to make sure I didn’t miss anything.